src/Evolution/Systems/Cce/BoundaryData.hpp

```diff
@@ -0,0 +1,224 @@
+ * Section 4.3 of \cite Bishop1998uk. The most direct comparison is to be made
+ * with equation (31) of \cite Barkett2019uae, which gives the null metric
+ * components explicitly. The time derivative is then (using notation from
+ * equation (31)  of \cite Barkett2019uae):
+ *
+ * \f{align}{
+ * \partial_{\bar u} g_{\bar u \bar \lambda} =
+ * \partial_{\bar u} g_{\bar \lambda \bar \lambda} =
+ * \partial_{\bar u} g_{\bar \lambda \bar A} &= 0 \\
+ * \partial_{\bar u} g_{\bar u \bar u} &=
+ * \partial_{\breve t} g_{\breve t \breve t} \\
+ * \partial_{\bar u} g_{\bar u \bar A} &=
+ * \frac{\partial \breve x^{\breve i}}{\partial \bar x^{\bar A}}\\
+ * g_{\breve i \breve t}
+ * \partial_{\bar u} g_{\bar A \bar B}
+ * &= \frac{\partial \breve x^{\breve i}}{\partial \bar x^{\bar A}}
+ * \frac{\partial \breve x^{\breve j}}{\partial \bar x^{\bar B}}
+ * g_{\breve i \breve j}
+ * \f}
+ */
+void null_metric_and_derivative(
+    gsl::not_null<tnsr::aa<DataVector, 3, Frame::RadialNull>*> du_null_metric,
+    gsl::not_null<tnsr::aa<DataVector, 3, Frame::RadialNull>*> null_metric,
+    const jacobian_tensor& cartesian_to_spherical_jacobian,
+    const tnsr::aa<DataVector, 3>& dt_spacetime_metric,
+    const tnsr::aa<DataVector, 3>& spacetime_metric) noexcept;
+
+/*!
+ * \brief Computes the spatial unit normal vector \f$s^i\f$ to the spherical
+ * worldtube surface and its first time derivative.
+ *
+ * \details Refer to equation (20) of \cite Barkett2019uae for the expression of
+ * the spatial unit normal vector, and equation (23) of \cite Barkett2019uae for
+ * the first time derivative. Refer to \cite Bishop1998uk for more exposition
+ * about the overall construction of the coordinate transformations used for the
+ * intermediate null coordinates.
+ */
+void worldtube_normal_and_derivatives(
+    gsl::not_null<tnsr::I<DataVector, 3>*> worldtube_normal,
+    gsl::not_null<tnsr::I<DataVector, 3>*> dt_worldtube_normal,
+    const Scalar<DataVector>& cos_phi, const Scalar<DataVector>& cos_theta,
+    const tnsr::aa<DataVector, 3>& spacetime_metric,
+    const tnsr::aa<DataVector, 3>& dt_spacetime_metric,
+    const Scalar<DataVector>& sin_phi, const Scalar<DataVector>& sin_theta,
+    const tnsr::II<DataVector, 3>& inverse_spatial_metric) noexcept;
+
+/*!
+ * \brief Computes the null 4-vector \f$l^\mu\f$ on the worldtube surface that
+ * is to be used as the CCE hypersurface generator, and the first time
+ * derivative \f$\partial_u l^\mu\f$.
+ *
+ * \details For mathematical description of our choice of the null generator,
+ * refer to equation (22) of \cite Barkett2019uae, and for the first time
+ * derivative see (25) of \cite Barkett2019uae.  Refer to \cite Bishop1998uk for
+ * more exposition about the overall construction of the coordinate
+ * transformations used for the intermediate null coordinates.
+ */
+void null_vector_l_and_derivatives(
+    gsl::not_null<tnsr::A<DataVector, 3>*> du_null_l,
```

Comment:
Should these `null` quantities be in the `RadialNull` frame? (I don't know about worldtube_normal)